multiplication_models/functions_for_all_nutrients.py

```diff
@@ -0,0 +1,158 @@
+    return data_frame
+
+
+def generate_overall_coverage_rates(filepath,
+                                    nutrient,
+                                    vehicle,
+                                    coverage_levels,
+                                    years,
+                                    location_ids,
+                                    subpopulations):
+    """This function generates baseline and counterfactual coverage rates of fortification for a specified
+    nutrient and vehicle pair. The baseline coverage rates are assumed to remain constant from over all of
+    the specified years. The alternative coverage rates are assumed to jump from the baseline rate in the
+    first specified year to either 20/50/80 percent (or the defined coverage levels) of the proportion
+    of the population consuming fortifiable/industrially produced vehicle in the second year of the specified
+    years and then remains constant for the remaining years.
+
+    The location_ids paramter can be set to 'all' or a list of specific location IDs.
+
+    The subpopulation parameter should be carefully evaluated and set to a list of the desired subpopulations
+    of interest that may be specific to women of reproductive age or children under 5, etc.
+    """
+
+    data = pd.read_csv(filepath)
+    if location_ids == 'all':
+        data = (data.loc[data.sub_population.isin(subpopulations)].drop_duplicates())
+    else:
+        data = (data.loc[data.location_id.isin(location_ids)]
+            .loc[data.sub_population.isin(subpopulations)].drop_duplicates())
+
+    # the following is a transformation for a potential data issue and should be removed when resolved
+    data['value_mean'] = data['value_mean'].replace(100, 100 - 0.00001 * 2)
+    data['value_025_percentile'] = data['value_025_percentile'].replace(100, 100 - 0.00001 * 3)
+    data['value_975_percentile'] = data['value_975_percentile'].replace(100, 100 - 0.00001)
+
+    data = data.loc[data.vehicle == vehicle].loc[data.nutrient.isin([nutrient, 'na'])]
+    data['value_std'] = (data.value_975_percentile - data.value_mean) / 1.96
+    data['a'] = (0 - data.value_mean) / data.value_std
+    data['b'] = (100 - data.value_mean) / data.value_std
+
+    cov_a = data.loc[data.value_description == 'percent of population eating fortified vehicle'].drop(
+        columns='value_description')
+    cov_b = data.loc[data.value_description == 'percent of population eating industrially produced vehicle'].drop(
+        columns='value_description')
+    cov_a = generate_coverage_parameter_draws(cov_a)
+    cov_b = generate_coverage_parameter_draws(cov_b)
+
+    assert np.all(cov_a <= cov_b), "Error: coverage parameters are not logically ordered"
+
+    baseline_coverage = pd.DataFrame()
+    for year in years:
+        temp = cov_a.copy()
+        temp['year'] = year
+        baseline_coverage = pd.concat([baseline_coverage, temp])

```

Comment:
I have a theory that it is cooler to make a list of DataFrames and then pd.concat the whole list, instead of repeatedly concating in a loop.  Writing this comment has made me realize that I have not tested my theory, though.